### ISSUE TITLE
Rule proposal: Named object destructuring

### ISSUE THREAD

**Issue by Richienb** [sindresorhus/eslint-plugin-unicorn#2401]
### Description

Readability/consistency

It took me a moment to figure out what this line does:
https://github.com/sindresorhus/open/blob/2ea66da8e8b20880d235447cf4c94ba275da6a5a/index.js#L307

### Fail

```js
const {[a]: b} = c;
```

I think we should still detect it here since I believe consistency should be sacrificed for readability (does someone remember what that 3-step or so style guide was called?).
```js
const {d} = g;
const {[a]: b} = c;
const {f} = h;
```

### Pass

```js
const b = c[a];
```

```js
const {d} = g;
const b = c[a];
const {f} = h;
```

### Proposed rule name

no-unreadable-object-destructuring

**Comment by fregante** [comment]
I agree, IMHO only very basic destructuring should ever be allowed, anything else is way harder to parse than the basic version.

**Comment by sindresorhus** [comment]
👍 I think it would make sense to add a `no-unreadable-object-destructuring` rule.

**Comment by sindresorhus** [comment]
Would be good to try to think of other anti-patterns that this rule could also catch.

**Comment by fisker** [comment]
The computed destructuring is readable to me, but people may get confused by some of the following cases 

1. Renamed 

	```js
	const  {a: renamed} = b
	```

2. Default values

	```js
	const {a = b} =c
	```

3. Deep destructuring 

	```js
	const {
		a: {
			b
		}
	} = d
	```

4. Array destructuring inside object destructuring 


	```js
	const {
		a: [b]
	} = d
	```

--- 

Combined all cases above

```js
const {
	[a]: {
		[b]: c = d,
		e: [f]
	} = {}
} = g
```

**Comment by fregante** [comment]
3 and 4 are visually same, one is `{b}` and the other one is `[b]`, neither one is readable 👍 

Renames and default values are not _super_ common, but they're very useful when setting defaults from options objects:

```js
const {
  enabled = true,
  id,
  depth = 1,
  replacements = []
} = options;
```

Maybe combining both still isn't ideal:


```js
const {
  enabled: isEnabled = true,
  id: name,
  depth = 1,
  replacements: array = []
} = options;
```

Combining it with inline types is particularly lethal, but that's another issue.

---

I suppose this rule should also apply to function parameters:

```js
function foo({
	[a]: {
		[b]: c = d,
		e: [f]
	} = {}
}) {}
```

**Comment by sindresorhus** [comment]
Renames and default values are very useful and should be allowed.

Deep restructuring should be allowed to a certain level. Maybe two levels down.

Array destructuring inside an object is pretty unreadable and should not be allowed.

**Comment by fisker** [comment]
When deep destructuring and default value can also confusing when the right side has multiple properties.

<details>
<summary>Shorthand</summary>



```js
const {
	a = {
		b,
		c
	}
} = d
```

```js
const {
	a: {
		b,
		c
	}
} = d
```

```js
const {
	a: e = {
		b,
		c
	}
} = d
```

</details>



<details>
<summary>Longhand</summary>



```js
const {
	a = {
		b: e,
		c
	}
} = d
```

```js
const {
	a: {
		b: e,
		c
	}
} = d
```

```js
const {
	a: {
		b = e,
		c
	}
} = d
```

</details>
